src/include/logging/checkpoint_manager.h

```diff
@@ -4,66 +4,147 @@
-//===--------------------------------------------------------------------===//
+namespace storage {
+  class Database;
+  class DataTable;
+  class TileGroup;
+  class TileGroupHeader;
+}
+
+namespace logging {
 
 class CheckpointManager {
- public:
+  // Deleted functions
   CheckpointManager(const CheckpointManager &) = delete;
   CheckpointManager &operator=(const CheckpointManager &) = delete;
   CheckpointManager(CheckpointManager &&) = delete;
-  CheckpointManager &operator=(CheckpointManager &&) = delete;
+  CheckpointManager &operator=(const CheckpointManager &&) = delete;
 
-  CheckpointManager() : is_running_(false) {}
 
+public:
+  CheckpointManager() :
+    is_running_(false),
+    checkpoint_interval_(DEFAULT_CHECKPOINT_INTERVAL) {
+    recovery_thread_count_ = 1;
+    // max_checkpointer_count_ = std::thread::hardware_concurrency() / 2;
+    max_checkpointer_count_ = 10;
+  }
   virtual ~CheckpointManager() {}
 
-  static CheckpointManager &GetInstance() {
-    static CheckpointManager checkpoint_manager;
-    return checkpoint_manager;
+  void SetCheckpointInterval(const int &checkpoint_interval) {
+    checkpoint_interval_ = checkpoint_interval;
+  }
+
+  void SetDirectories(const std::vector<std::string> &checkpoint_dirs) {
+    if (checkpoint_dirs.size() > 0) {
+      ckpt_pepoch_dir_ = checkpoint_dirs.at(0);
+    }
+    // check the existence of checkpoint directories.
+    // if not exists, then create the directory.
+    for (auto checkpoint_dir : checkpoint_dirs) {
+      if (LoggingUtil::CheckDirectoryExistence(checkpoint_dir.c_str()) == false) {
+        LOG_INFO("Checkpoint directory %s is not accessible or does not exist", checkpoint_dir.c_str());
+        bool res = LoggingUtil::CreateDirectory(checkpoint_dir.c_str(), 0700);
+        if (res == false) {
+          LOG_ERROR("Cannot create directory: %s", checkpoint_dir.c_str());
+        }
+      }
+    }
+
+    checkpoint_dirs_ = checkpoint_dirs;
+    checkpointer_count_ = checkpoint_dirs_.size();
   }
 
-  virtual void Reset() { is_running_ = false; }
+  void SetRecoveryThreadCount(const size_t &recovery_thread_count) {
+    recovery_thread_count_ = recovery_thread_count;
+    if (recovery_thread_count_ > max_checkpointer_count_) {
+      LOG_ERROR("# recovery thread cannot be larger than # max checkpointer");
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  virtual void StartCheckpointing();
+  virtual void StopCheckpointing();
+
+  size_t DoRecovery();
+
+protected:
+  std::string GetCheckpointFileFullPath(size_t checkpointer_id, size_t virtual_checkpointer_id, oid_t database_idx, oid_t table_idx, size_t epoch_id) {
+    return checkpoint_dirs_.at(checkpointer_id) + "/" + checkpoint_filename_prefix_ + "_" + std::to_string(virtual_checkpointer_id) + "_" + std::to_string(database_idx) + "_" + std::to_string(table_idx) + "_" + std::to_string(epoch_id);
+  }
+
+  // recover logic
+  size_t RecoverPepoch(std::vector<std::vector<size_t>> &database_structures);
```

Comment:
There are a lot of important functions here but **none** of them are documented with what they do.